chore: Add GitHub action to publish plugin-registry latest

vinokurig opened the pull request:
Signed-off-by: Igor Vinokur <ivinokur@redhat.com>

<!-- Please review the following before submitting a PR:
Che's Contributing Guide: https://github.com/eclipse/che/blob/master/CONTRIBUTING.md
Pull Request Policy: https://github.com/eclipse/che/wiki/Development-Workflow#pull-requests

-->

### What does this PR do?
Publish the plugin-registry with `latest` version to `surge.sh` and `github-pages`. The action is triggered on each commit to the `crw-2-thel-8` branch. This is needed to generate devfile v2 templates for CRW, see https://issues.redhat.com/browse/CRW-2500.
### What issues does this PR fix or reference?
https://issues.redhat.com/browse/CRW-2644
<!-- #### Changelog -->
<!-- The changelog will be pulled from the PR's title. 
     Please provide a clear and meaningful title to the PR and don't include issue number -->


#### Release Notes
<!-- markdown to be included in marketing announcement - N/A for bugs -->


#### Docs PR (if applicable)
<!-- Please add a matching PR to [the docs repo](https://gitlab.cee.redhat.com/red-hat-developers-documentation/red-hat-devtools) and link that PR to this issue.
Both will be merged at the same time. -->

svor commented on .github/workflows/plugin-registry-latest-build-publish-content.yaml:
```diff
@@ -0,0 +1,86 @@
+          git diff-index --quiet HEAD || git commit -m "publish registry $VERSION_DIR - $(date)" -s
+
+          git push "https://$GITHUB_ACTOR:${{ secrets.GITHUB_TOKEN }}@github.com/$GITHUB_REPOSITORY.git" gh-pages
+          echo "Checkout the published plugin registry at https://$GITHUB_ACTOR.github.io/$GITHUB_REPO_NAME/$VERSION_DIR/v3/plugins/"
```
maybe it makes sense to add some information about plugin-registry into the URL, like:
`https://$GITHUB_ACTOR.github.io/$GITHUB_REPO_NAME/dependencies/che-plugin-registry/$VERSION_DIR/v3/plugins/`

WDYT?

nickboldt replied:
whether you include `dependencies/che-plugin-registry` in the path or not, doesn't really matter.

But what IS mandatory is the arch. Otherwise we have the same registry for BOTH x64 (openjdk) and ppc64le/s390x (openj9). 

Pleae add a path segment for `/$(uname -m)/`

nickboldt replied:
this is in the original spec in https://issues.redhat.com/browse/CRW-2644:
```
use a URL like https://redhat-developer.github.io/codeready-workspaces/dependencies/che-plugin-registry/[branchOrTag]/[ARCH]/v3/
```

vinokurig replied:
The URL is now looks like: 
```
https://redhat-developer.github.io/codeready-workspaces/che-plugin-registry/[branch]/[ARCH]/v3/
```
I don't think we need to include the `dependencies` sub folder.

nickboldt replied:
:+1:

nickboldt commented:
Do we need to publish to BOTH surge and github pages? If only one URL is going to be used in builds, why publish 2x?

nickboldt reviewed (changes requested):
needs arch support; unclear why we include BOTH surge and gh pages. Suggest only using gh.

nickboldt commented on .github/workflows/plugin-registry-latest-build-publish-content.yaml:
```diff
@@ -0,0 +1,86 @@
+#
+# Copyright (c) 2020-2022 Red Hat, Inc.
```
You guys and your backdated copyrights, I tell ya. New file = 2022. Old file = old date - 2022.

nickboldt commented on .github/workflows/plugin-registry-latest-build-publish-content.yaml:
```diff
@@ -0,0 +1,86 @@
+on:
+  push:
+    branches:
+      - crw-2-rhel-8
```
should we also be publishing for the stable branches, crw-2.*-rhel-8? But those should go to :latest instead of :next.

vinokurig replied:
done

vinokurig replied:
added generation of `next`

nickboldt commented on .github/workflows/plugin-registry-latest-build-publish-content.yaml:
```diff
@@ -0,0 +1,86 @@
+          node-version: '12'
+      - name: extract content
+        run: |
+          docker pull quay.io/crw/pluginregistry-rhel8:latest
```
if publishing from crw-2-rhel-8 branch (aka 2.x aka main) then use :next, not :latest

vinokurig replied:
reworked to pull the `2.x` tag

nickboldt replied:
2.x = CI / :next / future / crw-2-rhel-8.

2.yy = stable / :latest / upcoming release / crw-2.yy-rhel-8

nickboldt commented on .github/workflows/plugin-registry-latest-build-publish-content.yaml:
```diff
@@ -0,0 +1,86 @@
+          BUILD_DIR="/tmp/crw-plugin-registry/content"
+          GH_PAGES_DIR="$PWD/gh-pages"
+          GITHUB_REPO_NAME="codeready-workspaces"
+          VERSION_DIR="latest"
```
if publishing from crw-2-rhel-8 (2.x / main) branch, use :next, not :latest.

:latest is reserved for the stable branch, built from crw-2.yy-rhel-8 (currently crw-2.14-rhel-8).

vinokurig replied:
reworked to pull the `2.x` tag

vinokurig commented:
@nickboldt
> needs arch support; unclear why we include BOTH surge and gh pages. Suggest only using gh.

Added arch support, removed the `surge.sh` logic.

nickboldt commented on .github/workflows/plugin-registry-build-publish-content-gh-pages.yaml:
```diff
@@ -0,0 +1,82 @@
+          docker manifest inspect quay.io/crw/pluginregistry-rhel8:latest | jq -r '.manifests[] | .platform.architecture' \
+          > /tmp/crw-plugin-registry/architectures.json
+          for arch in $(cat /tmp/crw-plugin-registry/architectures.json); do
+            version="${GITHUB_REF_NAME:4:4}"
```
might be safer to just read https://github.com/redhat-developer/codeready-workspaces/blob/crw-2-rhel-8/dependencies/VERSION = 2.16 or https://github.com/redhat-developer/codeready-workspaces/blob/crw-2.15-rhel-8/dependencies/VERSION = 2.15

vinokurig replied:
what if another branch triggered the build?

nickboldt replied:
well...
if triggered from crw-2 branch, you want to build new :2.16 / :next bits.
if triggered from crw-2.15-rhel-8 branch, you want to build new :2.15 / :latest bits.

so in either case, we get the correct source image, resulting GH page + path.

vinokurig replied:
done

nickboldt commented on .github/workflows/plugin-registry-build-publish-content-gh-pages.yaml:
```diff
@@ -0,0 +1,82 @@
+          git diff-index --quiet HEAD || git commit -m "publish registry $VERSION - $(date)" -s
+          git push "https://$GITHUB_ACTOR:${{ secrets.GITHUB_TOKEN }}@github.com/$GITHUB_REPOSITORY.git" gh-pages
+          for arch in $(cat /tmp/crw-plugin-registry/architectures.json); do
+            echo "Checkout the published plugin registry at https://$GITHUB_ACTOR.github.io/$GITHUB_REPO_NAME/$VERSION_DIR/$arch/v3/plugins/"
```
"checkout" implies a git checkout... do you mean..
`echo "Plugin registry published to https://..."` ?

nickboldt commented on .github/workflows/plugin-registry-build-publish-content-gh-pages.yaml:
```diff
@@ -0,0 +1,82 @@
+on:
+  push:
+    branches:
+      - 'crw-[0-9].[0-9]+-rhel-8'
```
this will only fire for 2.yy branches, but not the 2.next branch, `crw-2-rhel-8`. 

Is the plan to have this only work for the stable branches, and have next builds broken until we are ready to branch?

vinokurig replied:
added generation of `next`

nickboldt commented on .github/workflows/plugin-registry-build-publish-content-gh-pages.yaml:
```diff
@@ -0,0 +1,82 @@
+          for arch in $(cat /tmp/crw-plugin-registry/architectures.json); do
+            version="${GITHUB_REF_NAME:4:4}"
+            docker pull quay.io/crw/pluginregistry-rhel8:"$version" --platform="$arch"
+            docker create --name pluginRegistry quay.io/crw/pluginregistry-rhel8:"$version" sh
```
Tried this locally and it fails when run on x64 against a ppc64le image:

```
$➔ arch=ppc64le
$➔ version=2.14
$➔ docker pull quay.io/crw/pluginregistry-rhel8:"$version" --platform="$arch"
2.14: Pulling from crw/pluginregistry-rhel8
Digest: sha256:650a715a08927b11d78d8f520d0d9b623a3f9193eb98e1aed6eeebcaf4517b15
Status: Downloaded newer image for quay.io/crw/pluginregistry-rhel8:2.14

$➔             docker create --name pluginRegistry quay.io/crw/pluginregistry-rhel8:"$version" sh
WARNING: The requested image's platform (linux/ppc64le) does not match the detected host platform (linux/amd64) and no specific platform was requested
461e496615e1ec669d2a97433cd455cfaeb40db50e548c21faaa9e31512f5f4e
```

is the ubuntu runner multiarch or smart enough to fake it? I don't see you using buildx here...

vinokurig replied:
fixed by adding `platform` param to the `create` command

nickboldt replied:
aha! nice.

https://vinokurig.github.io/codeready-workspaces/che-plugin-registry/crw-2.15-rhel-8/ppc64le/v3/plugins/redhat/vscode-apache-camel/latest/ shows we have reference to the correct `plugin-java11-openj9-rhel8` image, so this appears to be working :D

vinokurig commented on .github/workflows/plugin-registry-build-publish-content-gh-pages.yaml:
```diff
@@ -0,0 +1,82 @@
+          GH_PAGES_DIR="$PWD/gh-pages"
+          GITHUB_REPO_NAME="codeready-workspaces"
+          VERSION_DIR="che-plugin-registry"/"$GITHUB_REF_NAME"
+          LATEST_VERSION_DIR="che-plugin-registry/latest"
```
This builds branches like `crw-2.15-rhel-8 ...`(stable)

nickboldt replied:
yes, and it ALSO needs to build from the `crw-2-rhel-8` branch / images (CI)

vinokurig replied:
done

nickboldt commented on .github/workflows/plugin-registry-build-publish-content-gh-pages.yaml:
```diff
@@ -23,11 +24,19 @@ jobs:
           touch /tmp/crw-plugin-registry/architectures.json
-          version="${GITHUB_REF_NAME:4:4}"
+          if [[ "$GITHUB_REF_NAME" == "crw-2-rhel-8" ]]; then
+             version="latest"
```
why don't you want a 2.16 version here?

vinokurig replied:
Do you mean a version from `dependencies/VERSION`? For that we generate from stable branches.

nickboldt commented on .github/workflows/plugin-registry-build-publish-content-gh-pages.yaml:
```diff
@@ -42,8 +51,14 @@ jobs:
-          LATEST_VERSION_DIR="che-plugin-registry/latest"
+          VERSION=$(cat dependencies/VERSION)
+          if [[ "$GITHUB_REF_NAME" == "crw-2-rhel-8" ]]; then
+            VERSION_DIR="che-plugin-registry/next"
```
so... for ci/next, we publish to ONLY next, but for stable/latest we publish to BOTH 2.15 and latest. 

Do... we need difference?

vinokurig replied:
As far as I understand `next` is a development version and is always changing, while `stable branch / latest` is a fixed version which is not expected to change. That's why when `next` is updated we publish only it, but when `2.15` is updated it means that we need to publish the brunch and `latest` as a reference to the latest stable brunch.

nickboldt commented on .github/workflows/plugin-registry-build-publish-content-gh-pages.yaml:
```diff
@@ -59,7 +74,7 @@ jobs:
           done
           # Check if the branch is latest.
-          if ! [[ $(git ls-remote origin crw-2.$(( ${GITHUB_REF_NAME:6:2} + 1 ))-rhel-8) ]]; then
+          if [[ "$GITHUB_REF_NAME" != "crw-2-rhel-8" ]] && ! [[ $(git ls-remote origin crw-$(echo $VERSION + 0.01 | bc -l)-rhel-8) ]]; then
```
what `bc` is this? On my machine it's Beyond Compare... but I suspect you mean something else.

nickboldt replied:
are you just checking that a future stable branch exists (eg., for `GITHUB_REF_NAME` = crw-2.15-rhel-8, you appear to be testing if crw-2.16-rhel-8 is NOT a branch. 

Why?

nickboldt replied:
if the 2.15 branch **isn't** the latest, and this fires for a new commit in an older 2.yy branch (like 2.14)... isn't that a good thing?

vinokurig replied:
> what bc is this? On my machine it's Beyond Compare... but I suspect you mean something else.

This is [`Bash Calculator`](https://www.tutorialsandyou.com/bash-shell-scripting/bash-bc-18.html)
If just `$(( 2.15 + 0.01 ))` then it returns `2.0099999999999998`
> Are you just checking that a future stable branch exists (eg., for GITHUB_REF_NAME = crw-2.15-rhel-8, you appear to be testing if crw-2.16-rhel-8 is NOT a branch) ?
>
>Why?
>
> if the 2.15 branch isn't the latest, and this fires for a new commit in an older 2.yy branch (like 2.14)... isn't that a good thing?

This is needed to be sure that the triggered branch is latest. Otherwise we mustn't update the `latest` folder.

svor reviewed (approved):
LGTM

nickboldt commented:
gh-page branch set up so only the bot can publish there.

Now running https://github.com/redhat-developer/codeready-workspaces/actions/workflows/plugin-registry-build-publish-content-gh-pages.yaml for both 2.15 and 2.16 branches
